MapboxNavigation/ImageCache.swift

```diff
@@ -0,0 +1,153 @@
+
+                do {
+                    try data?.write(to: cacheURL)
+                } catch {
+                    NSLog("================> Failed to write data to URL \(cacheURL)")
+                }
+                dispatchCompletion()
+            }
+        } else {
+            dispatchCompletion()
+        }
+    }
+
+    private func storeImageInMemoryCache(_ image: UIImage, forKey key: String) {
+        memoryCache.setObject(image, forKey: key as NSString, cost: cacheCostForImage(image))
+    }
+
+    private func cachePathWithKey(_ key: String) -> String {
+        let cacheKey = cacheKeyForKey(key)
+        return cacheURLWithKey(cacheKey).absoluteString
+    }
+
+    private func cacheURLWithKey(_ key: String) -> URL {
+        let cacheKey = cacheKeyForKey(key)
+        return diskCacheURL.appendingPathComponent(cacheKey)
+    }
+
+    func imageFromCache(forKey key: String?) -> UIImage? {
+        if let image = imageFromMemoryCache(forKey: key) {
+            return image
+        }
+
+        if let image = imageFromDiskCache(forKey: key) {
+            //TODO: test and extract helper method
+            storeImageInMemoryCache(image, forKey: key!)
+            return image
+        }
+
+        return nil
+    }
+
+    func clearMemory() {
+        memoryCache.removeAllObjects()
+    }
+
+    func clearDisk(completion: NoArgBlock?) {
+        self.diskAccessQueue.async {
+            do {
+                try self.fileManager!.removeItem(at: self.diskCacheURL)
+            } catch {
+                NSLog("================> Failed to remove cache dir: \(self.diskCacheURL)")
+            }
+
+            self.createCacheDirIfNeeded()
+
+            if let completion = completion {
+                DispatchQueue.main.async {
+                    completion()
+                }
+            }
+        }
+    }
+
+    private func cacheKeyForKey(_ key: String) -> String {
+        if let keyAsURL = URL(string: key) {
+            return keyAsURL.lastPathComponent
+        }
+        return key
+    }
+
+    private func imageFromMemoryCache(forKey key: String?) -> UIImage? {
+        guard let key = cacheKeyForKey(key!) as NSString! else {
+            return nil
+        }
+        return memoryCache.object(forKey: key)
+    }
+
+    private func imageFromDiskCache(forKey key: String?) -> UIImage? {
+        do {
+            let data = try Data.init(contentsOf: self.cacheURLWithKey(key!))
+            return UIImage.init(data: data)
+        } catch {
+            NSLog("================> Failed to load data at URL: \(self.cacheURLWithKey(key!))")
+            return nil
+        }
+    }
+
+    private func cacheCostForImage(_ image: UIImage) -> Int {
+        return Int(image.size.height * image.size.width * image.scale * image.scale);
+    }
+
+    private func createCacheDirIfNeeded() {
+        if fileManager!.fileExists(atPath: diskCacheURL.absoluteString) == false {
```

Comment:
Can we gently unwrap `fileManager`  here?

Reply:
Done - also refactored to rely less on data from self in closures, addressing @frederoni 's previous observation about the transient retain cycles on the disk access queue.